### DISCUSSION TITLE
Json serialization property order

### DISCUSSION THREAD

**Discussion by goncalo-oliveira** [kubernetes-client/csharp#1222] [Ideas]
It would be nice to have properties ordered in the output of a serialization. Not the most important thing in the world, but it would be interesting.

A custom converter that orders all properties alphabetically would probably fix it. There are a few challenges there, such as `JsonExtensionDataAttribute`, but it is feasible.

Or... giving the possibility to extend the `JsonSerializerOptions` in the static `KubernetesJson` class.

As I said, not the most important thing in the world. To me it is a bit annoying to read a json dump and have the `apiVersion` and `kind` at the bottom.

**Comment by tg123** [comment]
hmm, it should be configurable. maybe something like https://github.com/kubernetes-client/csharp/blob/master/src/KubernetesClient.Models/ModelVersionConverter.cs

**Reply by goncalo-oliveira** [reply]
I think it's worth mentioning that the YAML serializer also has some weird symptoms. Maybe it should be a separate discussion...

Consider the following

```csharp
var app = new V1Alpha1Application
{
    ApiVersion = "apps.faactory.io/v1alpha1",
    Kind = "Application",
    Metadata = new k8s.Models.V1ObjectMeta
    {
        Name = name,
        NamespaceProperty = ns
    },
    Spec = new V1Alpha1Application.ApplicationSpec
    {
        Image = specImage,
        Port = specPort
    }
};

var yaml = k8s.KubernetesYaml.Serialize( app );
```

And the yaml...

```yaml
spec:
  image: gcr.io/google-samples/hello-app:1.0
  port: 8080
metadata:
  name: confident-hawking
apiVersion: apps.faactory.io/v1alpha1
kind: Application
```

**Reply by cdavernas** [reply]
@goncalo-oliveira Did you find a solution/workaround?
@tg123 Any suggestion to get proper ordering when using `KubernetesYaml`?

**Reply by tg123** [reply]
it is ok, same behavior if kubectl get -o yaml
there is not spec saying apiversion 1st, kind 2nd, ...

**Reply by cdavernas** [reply]
@tg123 I have to respectfully disagree. While the current approach is functional, it effectively prevents reusing client resources in scenarios such as ArgoCD. In those cases, resources are not applied directly to Kubernetes but are instead stored in a repository to enable human-readable auditing.

It’s unfortunate that we can’t reuse the client’s resources for this purpose, especially since — as far as I know — addressing it would be relatively straightforward by either annotating properties with a `YamlMember` attribute or implementing a contract resolver.
